datalad_next/runners/data_processors.py

```diff
@@ -0,0 +1,316 @@
+- split a stream of characters at line-ends
+- convert a line of text into JSON
+
+Data processors have a common interface and can be chained. For example,
+one can pass data chunks, where each chunk is a byte-string, into a chain
+of two data processors: a decode-processor that converts bytes into strings,
+and a linesplit-processor that converts character-streams into lines. The result
+of the chain would be lines of text.
+
+Data processors might need to buffer some data before yielding their result. The
+"driver" of the data processing chains supports the buffering of input data for
+individual processors. Therefore, most data processors do not need to store
+state themselves and can be quite simple
+"""
+from __future__ import annotations
+
+import json
+from collections import defaultdict
+from collections.abc import Generator
+from functools import partial
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    List,
+    Union,
+)
+
+from datalad_next.runners import LineSplitter
+
+
+# Protocol machines are callables that have the following signature:
+#
+#    def process(self, data: list[T], final: bool) -> tuple[list[N] | None, list[T]]:
+#       ...
+#
+# where N is the type that is returned by processor. The return value is a
+# consisting of optional results, i.e. list[N] | None, and a number of input
+# elements that were not processed and should be presented again, when more
+# data arrives from the "preceding" element.
+
+StrList = List[str]
+BytesList = List[bytes]
+StrOrBytes = Union[str, bytes]
+StrOrBytesList = List[StrOrBytes]
+
+
+def jsonline_processor(lines: StrOrBytesList,
+                       _: bool = False
+                       ) -> tuple[list[tuple[bool, Any]], StrOrBytesList]:
+    """
+    A processor that converts lines into JSON objects, if possible.
+
+    lines: StrOrBytesList
+      A list containing strings or byte-strings that that hold JSON-serialized
+      data.
+
+    Returns: tuple[list[Tuple[bool, StrOrBytes]], StrOrByteList]
+      The result, i.e. the first element of the result tuple, is a list that
+      contains one tuple for each element of `lines`. The first element of the
+      tuple is a bool that indicates whether the line could be converted. If it
+      was successfully converted the value is `True`. The second element is the
+      Python structure that resulted from the conversion if the first element
+      was `True`. If the first element is `False`, the second element contains
+      the input that could not be converted.
+    """
+    result = []
+    for line in lines:
+        assert len(line.splitlines()) == 1
+        try:
+            result.append((True, json.loads(line)))
+        except json.decoder.JSONDecodeError:
+            result.append((False, lines))
+    return result, []
+
+
+def splitlines_processor(data_chunks: StrOrBytesList,
+                         final: bool = False
+                         ) -> tuple[StrOrBytesList, StrOrBytesList]:
```

Comment:
What is the relationship to `SplitLinesProcessor`? That one is undocumented but used. This one is documented but unused outside the tests, it seems.

Reply:
`SplitLineProcessor` is gone.